### PR TITLE
Fix chain cover background update to work with split out event persisters

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes:

```
2021-01-14 16:41:53,667 - synapse.storage.background_updates - 120 - ERROR - background_updates-0 - Error doing update
Capture point (most recent call last):
  File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/synapse/src/synapse/app/homeserver.py", line 519, in <module>
    main()
  File "/home/synapse/src/synapse/app/homeserver.py", line 515, in main
    run(hs)
  File "/home/synapse/src/synapse/app/homeserver.py", line 506, in run
    logger=logger,
  File "/home/synapse/src/synapse/app/_base.py", line 134, in start_reactor
    run()
  File "/home/synapse/src/synapse/app/_base.py", line 118, in run
    run_command()
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/base.py", line 1283, in run
    self.mainLoop()
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/base.py", line 1292, in mainLoop
    self.runUntilCurrent()
  File "/home/synapse/src/synapse/metrics/__init__.py", line 548, in f
    ret = func(*args, **kwargs)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/base.py", line 886, in runUntilCurrent
    f(*a, **kw)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 501, in errback
    self._startRunCallbacks(fail)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 568, in _startRunCallbacks
    self._runCallbacks()
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 1475, in gotResult
    _inlineCallbacks(r, g, status)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
Traceback (most recent call last):
  File "/home/synapse/src/synapse/storage/background_updates.py", line 117, in run_background_updates
    self.BACKGROUND_UPDATE_DURATION_MS
  File "/home/synapse/src/synapse/storage/background_updates.py", line 227, in do_next_background_update
    await self._do_background_update(desired_duration_ms)
  File "/home/synapse/src/synapse/storage/background_updates.py", line 264, in _do_background_update
    items_updated = await update_handler(progress, batch_size)
  File "/home/synapse/src/synapse/storage/databases/main/events_bg_updates.py", line 843, in _chain_cover_index
    "_chain_cover_index", _calculate_auth_chain, last_depth, last_stream
  File "/home/synapse/src/synapse/storage/database.py", line 653, in runInteraction
    **kwargs,
  File "/home/synapse/src/synapse/storage/database.py", line 729, in runWithConnection
    self._db_pool.runWithConnection(inner_func, *args, **kwargs)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/enterprise/adbapi.py", line 306, in _runWithConnection
    compat.reraise(excValue, excTraceback)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/compat.py", line 464, in reraise
    raise exception.with_traceback(traceback)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/enterprise/adbapi.py", line 297, in _runWithConnection
    result = func(conn, *args, **kw)
  File "/home/synapse/src/synapse/storage/database.py", line 723, in inner_func
    return func(db_conn, *args, **kwargs)
  File "/home/synapse/src/synapse/storage/database.py", line 523, in new_transaction
    r = func(cursor, *args, **kwargs)
  File "/home/synapse/src/synapse/storage/databases/main/events_bg_updates.py", line 836, in _calculate_auth_chain
    self.hs.get_datastores().persist_events._add_chain_cover_index(
AttributeError: 'NoneType' object has no attribute '_add_chain_cover_index'
```

Introduced in #9029

### CODE REVIEW

File: synapse/storage/database.py [hunk 1]
```diff
@@ -412,6 +413,16 @@ def __init__(
                 self._check_safe_to_upsert,
             )
 
+        # We define this sequence here so that it can be referenced from both
+        # the DataStore and PersistEventStore.
```
Comment (clokep) [inline synapse/storage/database.py:417]: This feels awfully dirty to put this here. Should we have a `PersistEventWorkerStore` that contains just this?

Comment (erikjohnston) [reply synapse/storage/database.py:417]: It is dirty. I don't think having a shared base class works as we the `PersistEventStore` isn't part of the mixin, and so you'd lead to two instances of the sequence (which is fine for postgres, but not fine for SQLite)

**Review by clokep — APPROVED** [review]
✅ But I'm not thrilled about it. 😛

**Comment by erikjohnston** [comment]
>  But I'm not thrilled about it.

Neither am I :sob:

**Pull request merged**